### ISSUE TITLE
0.4.0-M3?

### ISSUE THREAD

**Issue by eed3si9n** [scala-native/scala-native#2043]
Scala Native 0.4.0-M2 came out in May, 2019. It's been over a year, and there's been much progress and bug fixes.
Could the maintainers publish the milestones more regularly, like every month?

I'm particularly interested in https://github.com/scala-native/scala-native/issues/1770 / https://github.com/scala-native/scala-native/pull/1805. I keep forgetting that `BigDecimal` is broken in 0.4.0-M2 and it should not be part of my build matrix.

**Comment by sjrd** [comment]
We should release a 0.4.0 final pretty soon, now. It's probably not worth publishing an M3 that will send the maintainers into a publishing frenzy just before asking them to do the same for 0.4.0.

**Comment by JD557** [comment]
If M3 is going to be skipped, I think it would still be nice to have a feature freeze some days before the actual release and ask mantainers to try the latest snapshot and report bugs that should be blockers.

I know that there's already a `blocker` label, but some important issues don't have such label (such as #1950) and I fear that, without some kind of prerelease, some issues like that might end up in 0.4.0, and then we end up with the same publishing frenzy problem, but for 0.4.1 😛

**Comment by sjrd** [comment]
Blocker = the fix would break backward binary compatibility. I don't think this is the case of the issue you mention.

**Comment by eed3si9n** [comment]
IMO having a regular pre-release of good quality is an ok problem to have especially if the final goal date is announced together. If we're down to 2 blocker PRs, could we merge them, release 0.4.0-RC1, and see if there are further blockers in 2 week fermentation? That way some library authors who wish to participate in the RC cycle would publish against 0.4.0-RC1 to try the glorious build matrix, while others could wait until the 0.4.0 final ships.

**Comment by catap** [comment]
@sjrd maybe start from `-M3`, after than `-RC1`, and so on?

Idea to release `0.4.0` after a year and half of development and near to double features looks like a way to garantue that `0.4.1` with bugfixes will be very soon :)

**Comment by sjrd** [comment]
The need for 0.4.1 very soon is not an issue, as long as it doesn't break backward binary compatibility. That is a fundamental difference between one more M or RC, versus a final.

And if we really need a binary breaking change, then we can release 0.5.0; it won't be harder to do that than to release another RC.

**Comment by catap** [comment]
@sjrd why rush? That is disadvantages to release `-RC1`?

**Comment by sjrd** [comment]
The disadvantage is that, if it ends up being fine, we lost a round of publishing by library authors. If it ends up requiring other binary breaking changes, we lost nothing.

**Comment by catap** [comment]
@sjrd if we start using `_sn0.4_2.11` instead of `_native0.4.0-M2_2.11` it will fix this issue, won't it?

**Comment by sjrd** [comment]
If we start using `_sn0.4_2.11` in an RC1, and we need a fix that breaks binary compat, we also need to bump to 0.5.0 to have a new `_sn0.5_2.11`. So we have gained nothing, again.

**Comment by eed3si9n** [comment]
All road eventually leads to setting up Community Build 😄

**Comment by catap** [comment]
@sjrd but if we release `-RC1` and setup a date when `-RC2` and final version is released... won't it create any issue?

**Comment by sjrd** [comment]
I don't understand how this helps.

**Comment by ekrich** [comment]
Hey guys and gals, I have been thinking about this allot and am now thinking that an `M3`, `M4`, `...`, `RC1` would be a good idea.

First of all I very much liked the approach used by Dotty where there were regular releases and I just followed along and updated and published my library. I followed this for about a year and it allowed me to adapt as things changed. There were many missing features and bugs but my bug reports and reproducers helped the end product and were used for test cases. Each time a new release seemed to fix a problem and allowed me to adapt little by little. 

Here are some of the benefits I came up with and I see no downside except the hassle and time spent on the release(s). The other danger is that we drag things out too long.

1. We haven't had a release in a long time and there are tons of improvements and features. This is a big deal.
2. This allows us to delay releasing a production release as we work out different issues and features.
3. Downstream projects have a real target to update and they will need to cross compile for `2.11-2.13` but they have time.
4. The project will have more time to add features that are needed that we may not want to attack otherwise.
5. There is less stress and we can get more people involved, interested, and excited.
6. We don't need to worry about breaking things and changing things as they are pre-releases.
7. With the JUnit support we should get proper testing support in Metals?

I have three downstream projects and a PR to `scala-collection-compat` that I would need to update to follow along but I can follow along or skip if I want or ignore until `0.4.0`. Some core projects would need to follow like `scala-collection-compat` and the test libraries and tools would need to follow for further downstream projects. If using JUnit there is no delay.

The other approach which to me is also attractive is just to button this one up as soon as possible but once we put the `0.4.0` stamp there are much more pressure for a really good product.

To me the delay can allow us to add 32bit, support for threads, real support for `structs` with field names and stack allocation. We can also make sure our `Zone` support is more solid and make sure our current macros will port over to Scala 3. We could also consider Scala 3 support if everything is settled down and we are pretty sure things are really solid. Certainly, any of these could be left out. If we go with this approach then `0.4.0` when it finally comes out could be really good. Realistically this approach would probably take most of next year with 2 to 3 more `M/RC` releases.

**Comment by neozenith** [comment]
Is there value in adding the issues and PRs still considered outstanding for v0.4.0 into this github milestone?

https://github.com/scala-native/scala-native/milestone/21

Or setting up new `v0.4.0-M3` and `v0.4.0-RC1` github milestones to make it clear to library authors what is still remaining per milestone/candidate release?

**Comment by scalavision** [comment]
My experience is similar to yours @ekrich , the last couple of months I have delved into Scala 3, finding bugs, both when learning from the docs, and when trying to convert open source projects to Dotty.

As of now, I am kind of waiting for the new release of Scala Native, as I don't want to learn a lot of things in `M2` that would no longer be valid in a final release of 0.4.0. With temporary releases of M's, I could follow along on the project, following migration notes. It is quite hard to follow along with changes currently, without temporary milestone releases.

But for all I know, the changes are only internal and not surfacing the end user. If that is the case, then the M's are probably not that helpful in the end. Also, if it puts a lot of pressure on the development, then I am not for it. I mean, feedback with increased number of issues could also take a lot of resources, that otherwise could have been spent on just fixing them ..

**Comment by ekrich** [comment]
As announced on Gitter, release of `0.4.0` will be soon 🎉 
:point_up: [January 12, 2021 3:17 AM](https://gitter.im/scala-native/scala-native?at=5ffd853dd5f4bf2965dfe84c)
